### PR TITLE
Preview precompiled letters sent using an API test key

### PR DESCRIPTION
**Pull request by kentsanggds**
## What 
To break down this PR, you can have a look at this first - 

https://github.com/alphagov/notifications-api/pull/1772

This PR will access the PDF from the correct bucket for a preview image sent using a test API key

## Dependencies 

This is a branch off https://github.com/alphagov/notifications-api/tree/ken-upload-test_key-precompiled-to-test_letters-bucket, so should be merged after that branch has merged or can be merged in place of the merging of the parent branch.

## Reference 

https://www.pivotaltracker.com/story/show/155974923

### CODE REVIEW

File: app/config.py [hunk 1]
```diff
@@ -395,6 +398,7 @@ class Staging(Config):
     NOTIFY_ENVIRONMENT = 'staging'
     CSV_UPLOAD_BUCKET_NAME = 'staging-notify-csv-upload'
     LETTERS_PDF_BUCKET_NAME = 'staging-letters-pdf'
+    TEST_LETTERS_BUCKET_NAME = 'staging-test-letters'
```
Comment (richardc0) [inline app/config.py]: We should probably prefix all of these with the environment at some point

File: app/letters/utils.py [hunk 2]
```diff
@@ -5,24 +5,32 @@
...
+PRECOMPILED_BUCKET_PREFIX = '{folder}NOTIFY.{reference}'
 
 
-def get_letter_pdf_filename(reference, crown):
-    now = datetime.utcnow()
+def get_folder_name(_now, is_test_letter):
+    if is_test_letter:
+        folder_name = ''
+    else:
+        print_datetime = _now
```
Comment (richardc0) [inline app/letters/utils.py]: Should we use this as an opportunity to use created_at instead of now()?

It would be less prone to error and would mean we haven't got to search the bucket but use the created_at date. Get file would be just obj = s3.Object(bucket_name=bucket_name, key=item.key)

Technically a notification could be created at <deadline>:999 and this be ran on <deadline>:001 which would cause the folder name to be different for the get_letter_pdf function

File: tests/app/celery/test_letters_pdf_tasks.py [hunk 3]
```diff
@@ -28,25 +28,6 @@ def test_should_have_decorated_tasks_functions():
     assert create_letters_pdf.__wrapped__.__name__ == 'create_letters_pdf'
 
 
-@pytest.mark.parametrize('crown_flag,expected_crown_text', [
```
Comment (richardc0) [inline tests/app/celery/test_letters_pdf_tasks.py]: Why were these deleted? They seem like valid tests (may need updating to determine if it was a test or not test though)

File: tests/app/letters/test_letter_utils.py [hunk 4]
```diff
@@ -13,6 +33,70 @@ def test_get_bucket_prefix_for_notification_valid_notification(sample_notificati
     ).upper()
 
 
+@freeze_time(FROZEN_DATE_TIME)
```
Comment (richardc0) [inline tests/app/letters/test_letter_utils.py]: I think that the freeze time masks the issue with using a transient date

File: app/config.py (hunk 1 above)
Comment (kentsanggds) [reply app/config.py]: Not sure what you mean here, they are already prefixed with the environment?

File: app/letters/utils.py (hunk 2 above)
Comment (kentsanggds) [reply app/letters/utils.py]: Wanted to keep this PR as small as possible, that other issue could be fixed in another PR

File: tests/app/celery/test_letters_pdf_tasks.py (hunk 3 above)
Comment (kentsanggds) [reply tests/app/celery/test_letters_pdf_tasks.py]: They weren't deleted, just moved to a more logical place to `test_letter_utils.py`

File: tests/app/letters/test_letter_utils.py (hunk 4 above)
Comment (kentsanggds) [reply tests/app/letters/test_letter_utils.py]: I would introduce a test specific for that, and maybe fix it in another PR

File: app/letters/utils.py (hunk 2 above)
Comment (kentsanggds) [reply app/letters/utils.py]: Had another thought on this, if the letter gets processed but then we create the pdf after the cut off time, there's a potential that it might never gets processed if it's using the `created_at` time to indicate which subfolder it will be placed in. Think we need a bit more discussion around this.
